api/src/nv_ingest_api/internal/extract/pdf/engines/pdfium.py

```diff
@@ -377,6 +378,94 @@ def _extract_page_elements(
     return extracted_page_elements
 
 
+# ------------------------------------------------------------------ #
+#  Render one page – now returns (page_idx, image, padding_offsets)  #
+# ------------------------------------------------------------------ #
+def _render_page_from_file(
+    pdf_source: Union[str, bytes],
+    page_idx: int,
+    size: Tuple[int, int],
+    execution_trace_log: Optional[List] = None,
+) -> Tuple[int, np.ndarray, Tuple[int, int]]:
+    """
+    Opens `pdf_source`, renders `page_idx`, and returns
+    (page_idx, image_array, padding_offsets).
+    """
+    doc = libpdfium.PdfDocument(pdf_source)
+    page = doc.get_page(page_idx)
+
+    arrays, pads = pdfium_pages_to_numpy(
+        [page],
+        scale_tuple=size,
+        padding_tuple=size,
+        trace_info=execution_trace_log,
+    )
+
+    page.close()
+    doc.close()
+    return page_idx, arrays[0], pads[0]
+
+
+# ------------------------------------------------------------------ #
+#  Render an entire PDF in parallel – keeps natural page order       #
+# ------------------------------------------------------------------ #
+def render_single_pdf_parallel(
+    pdf_source: Union[str, bytes],
+    size: Tuple[int, int],
+    max_workers: int = 8,
+    execution_trace_log: Optional[List] = None,
+) -> List[Tuple[np.ndarray, Tuple[int, int]]]:
+    """
+    Render all pages of `pdf_source` in parallel and return a list of
+    (image_array, padding_offsets) tuples in natural page order.
+    """
+    doc = libpdfium.PdfDocument(pdf_source)
+    page_count = len(doc)
+    doc.close()
+
+    max_workers = min(max_workers, page_count)
+    images: List[Tuple[np.ndarray, Tuple[int, int]]] = [None] * page_count
+
+    if execution_trace_log is not None:
+        for idx in range(page_count):
+            execution_trace_log[f"trace::entry::pdf_extraction::pdfium_pages_to_numpy_{idx}"] = datetime.now()
+
+    with ProcessPoolExecutor(max_workers=max_workers) as pool:
+        futs = [pool.submit(_render_page_from_file, pdf_source, idx, size) for idx in range(page_count)]
+        for fut in as_completed(futs):
+            idx, img, pad = fut.result()
+            images[idx] = (img, pad)
+            if execution_trace_log is not None:
+                execution_trace_log[f"trace::exit::pdf_extraction::pdfium_pages_to_numpy_{idx}"] = datetime.now()
+
+    return images
```

Comment:
Maybe not now but I could see this using a lot of memory holding all of the images in memory here. Maybe at some point we introduce a generator concept that writes each image after successful generation to disk or the next stage in the pipeline without having to accumulate them all here.

Reply:
>Another thing to consider here is that there does not appear to be a specific reason for this one-page-per-process approach, it could be blocks of pages.

I like this idea, we can explore a block of pages and see how that performs and that should help elevate some of the concerns. 

I think b/w that and only keeping base64 strings around (instead of the whole images) , we can reduce memory  per process a bunch.